GraphicsContextGL external image and sync lifecycle handling is incorrect

kkinnunen-apple opened the pull request:
#### 8d8fa5a3d62041865fecf0d2558692a72f143126
<pre>
GraphicsContextGL external image and sync lifecycle handling is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=271642">https://bugs.webkit.org/show_bug.cgi?id=271642</a>
<a href="https://rdar.apple.com/125350987">rdar://125350987</a>

Reviewed by Dan Glastonbury.

External images and sync objects had following problems:
 - created with synchronous calls
 - leaked when context was abandoned

Rename the concepts from EGLImage and EGLSync to ExternalImage and
ExternalSync, to highlight that these are not cross context objects,
rather objects bound to a specific GraphicsContextGL.

Rename the destroy function to delete to make the objects look similar
to normal GL objects.

Allocate client (WP) side names and send them to server (GPUP) side
during creation. This makes the creation asynchronous.

Destroy the underlying EGL objects when the real ANGLE context is
destroyed.

Generate the implementations, similar to other normal GraphicsContextGL
API calls.

Fix a bug in clientWaitExternalSyncWithFlush: it was asynchronous
instead of synchronous.

Due to the special handling of C++ function names beginning with
`create`, `createFoveation(...)` has been renamed to
`addFoveation(...)`.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::createAndBindCompositorTexture):
(WebCore::createAndBindCompositorBuffer):
(WebCore::makeEGLImageSource):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createExternalImage):
(WebCore::GraphicsContextGLANGLE::bindEGLImage):
(WebCore::GraphicsContextGLANGLE::deleteExternalImage):
(WebCore::GraphicsContextGLANGLE::createExternalSync):
(WebCore::GraphicsContextGLANGLE::deleteExternalSync):
(WebCore::GraphicsContextGLANGLE::clientWaitExternalSyncWithFlush):
(WebCore::GraphicsContextGLANGLE::createAndBindEGLImage): Deleted.
(WebCore::GraphicsContextGLANGLE::destroyEGLImage): Deleted.
(WebCore::GraphicsContextGLANGLE::createEGLSync): Deleted.
(WebCore::GraphicsContextGLANGLE::destroyEGLSync): Deleted.
(WebCore::GraphicsContextGLANGLE::clientWaitEGLSyncWithFlush): Deleted.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLCocoa::createExternalImage):
(WebCore::GraphicsContextGLCocoa::bindEGLImage):
(WebCore::GraphicsContextGLCocoa::createExternalSync):
(WebCore::GraphicsContextGLCocoa::insertFinishedSignalOrInvoke):
(WebCore::GraphicsContextGLCocoa::createAndBindEGLImage): Deleted.
(WebCore::GraphicsContextGLCocoa::createEGLSync): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::createAndBindEGLImage): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::createEGLSync): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createBuffer):
(createFramebuffer):
(createProgram):
(createRenderbuffer):
(createShader):
(createTexture):
(depthRange):
(createVertexArray):
(createQuery):
(createSampler):
(createTransformFeedback):
(createQueryEXT):
(createExternalImage):
(deleteExternalImage):
(bindEGLImage):
(createEGLSync):
(deleteExternalSync):
(clientWaitExternalSyncWithFlush):
(destroyEGLImage): Deleted.
(destroyEGLSync): Deleted.
(clientWaitEGLSyncWithFlush): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createAndBindEGLImage): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::createEGLSync): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createExternalImage):
(WebKit::RemoteGraphicsContextGLProxy::deleteExternalImage):
(WebKit::RemoteGraphicsContextGLProxy::bindEGLImage):
(WebKit::RemoteGraphicsContextGLProxy::createEGLSync):
(WebKit::RemoteGraphicsContextGLProxy::deleteExternalSync):
(WebKit::RemoteGraphicsContextGLProxy::clientWaitExternalSyncWithFlush):
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLImage): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLSync): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::clientWaitEGLSyncWithFlush): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:
* Tools/Scripts/generate-gpup-webgl:
(cpp_type.is_reference):
(cpp_type):
(webkit_ipc_convert_expr):
(get_delete_func_object_name):
(webkit_ipc_msg.__init__):
(webkit_ipc_msg.__str__):
(context_proxy_cpp_webkit_ipc_generator.generate):
(webkit_ipc_cpp_impl.process_in_arg):
(webkit_ipc_cpp_impl.process_return_value):
(main):

Canonical link: <a href="https://commits.webkit.org/277323@main">https://commits.webkit.org/277323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de0b58c22d82db89b4a0900fed116031f60083b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47275 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49959 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43324 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49582 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31889 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38499 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47856 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40743 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19813 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/47139 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41905 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5319 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43635 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42308 "Found 1 new test failure: inspector/cpu-profiler/tracking.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51834 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22305 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18652 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45793 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23580 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44808 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10428 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24364 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23298 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -616,15 +624,19 @@ static bool needsEAGLOnMac()
-    if (!EGL_QueryDisplayAttribEXT(platformDisplay(), EGL_DEVICE_EXT, reinterpret_cast<EGLAttrib*>(&eglDevice)))
-        return nullptr;
+    if (!EGL_QueryDisplayAttribEXT(platformDisplay(), EGL_DEVICE_EXT, reinterpret_cast<EGLAttrib*>(&eglDevice))) {
+        addError(GCGLErrorCode::InvalidOperation);
```
👍

djg commented on Tools/Scripts/generate-gpup-webgl:
```diff
@@ -102,9 +102,9 @@ messages -> RemoteGraphicsContextGL NotRefCounted Stream {{
     void EnsureExtensionEnabled(String extension)
     void GetErrors() -> (GCGLErrorCodeSet returnValue) Synchronous
-    void DrawSurfaceBufferToImageBuffer(WebCore::GraphicsContextGL::SurfaceBuffer buffer, WebCore::RenderingResourceIdentifier imageBuffer) -> () Synchronous
+    void DrawSurfaceBufferToImageBuffer(enum:bool WebCore::GraphicsContextGLSurfaceBuffer buffer, WebCore::RenderingResourceIdentifier imageBuffer) -> () Synchronous
```
What is `enum:bool`?

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
GraphicsContextGL external image and sync lifecycle handling is incorrect
https://bugs.webkit.org/show_bug.cgi?id=271642
rdar://125350987

Reviewed by Dan Glastonbury.

External images and sync objects had following problems:
 - created with synchronous calls
 - leaked when context was abandoned

Rename the concepts from EGLImage and EGLSync to ExternalImage and
ExternalSync, to highlight that these are not cross context objects,
rather objects bound to a specific GraphicsContextGL.

Rename the destroy function to delete to make the objects look similar
to normal GL objects.

Allocate client (WP) side names and send them to server (GPUP) side
during creation. This makes the creation asynchronous.

Destroy the underlying EGL objects when the real ANGLE context is
destroyed.

Generate the implementations, similar to other normal GraphicsContextGL
API calls.

Fix a bug in clientWaitExternalSyncWithFlush: it was asynchronous
instead of synchronous.

Due to the special handling of C++ function names beginning with
`create`, `createFoveation(...)` has been renamed to
`addFoveation(...)`.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::createAndBindCompositorTexture):
(WebCore::createAndBindCompositorBuffer):
(WebCore::makeEGLImageSource):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createExternalImage):
(WebCore::GraphicsContextGLANGLE::bindEGLImage):
(WebCore::GraphicsContextGLANGLE::deleteExternalImage):
(WebCore::GraphicsContextGLANGLE::createExternalSync):
(WebCore::GraphicsContextGLANGLE::deleteExternalSync):
(WebCore::GraphicsContextGLANGLE::clientWaitExternalSyncWithFlush):
(WebCore::GraphicsContextGLANGLE::createAndBindEGLImage): Deleted.
(WebCore::GraphicsContextGLANGLE::destroyEGLImage): Deleted.
(WebCore::GraphicsContextGLANGLE::createEGLSync): Deleted.
(WebCore::GraphicsContextGLANGLE::destroyEGLSync): Deleted.
(WebCore::GraphicsContextGLANGLE::clientWaitEGLSyncWithFlush): Deleted.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLCocoa::createExternalImage):
(WebCore::GraphicsContextGLCocoa::bindEGLImage):
(WebCore::GraphicsContextGLCocoa::createExternalSync):
(WebCore::GraphicsContextGLCocoa::insertFinishedSignalOrInvoke):
(WebCore::GraphicsContextGLCocoa::createAndBindEGLImage): Deleted.
(WebCore::GraphicsContextGLCocoa::createEGLSync): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::createAndBindEGLImage): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::createEGLSync): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createBuffer):
(createFramebuffer):
(createProgram):
(createRenderbuffer):
(createShader):
(createTexture):
(depthRange):
(createVertexArray):
(createQuery):
(createSampler):
(createTransformFeedback):
(createQueryEXT):
(createExternalImage):
(deleteExternalImage):
(bindEGLImage):
(createEGLSync):
(deleteExternalSync):
(clientWaitExternalSyncWithFlush):
(destroyEGLImage): Deleted.
(destroyEGLSync): Deleted.
(clientWaitEGLSyncWithFlush): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createAndBindEGLImage): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::createEGLSync): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createExternalImage):
(WebKit::RemoteGraphicsContextGLProxy::deleteExternalImage):
(WebKit::RemoteGraphicsContextGLProxy::bindEGLImage):
(WebKit::RemoteGraphicsContextGLProxy::createEGLSync):
(WebKit::RemoteGraphicsContextGLProxy::deleteExternalSync):
(WebKit::RemoteGraphicsContextGLProxy::clientWaitExternalSyncWithFlush):
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLImage): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLSync): Deleted.
(WebKit::RemoteGraphicsContextGLProxy::clientWaitEGLSyncWithFlush): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:
* Tools/Scripts/generate-gpup-webgl:
(cpp_type.is_reference):
(cpp_type):
(webkit_ipc_convert_expr):
(get_delete_func_object_name):
(webkit_ipc_msg.__init__):
(webkit_ipc_msg.__str__):
(context_proxy_cpp_webkit_ipc_generator.generate):
(webkit_ipc_cpp_impl.process_in_arg):
(webkit_ipc_cpp_impl.process_return_value):
(main):

Canonical link: https://commits.webkit.org/277323@main
